### DISCUSSION TITLE
Generating type-safe accessors from nodeset XML?

### DISCUSSION THREAD

**Discussion by oliver** [open62541pp/open62541pp#169] [Q&A] [answered]
Hi,

I'm just getting started with OPC-UA, and I was wondering: is there a way to generate type-safe C++ accessor methods from a Nodeset XML?

I have the Nodeset XML file for my machine interface; and I can use it on server side (with the [XML Nodeset Compiler](https://www.open62541.org/doc/master/nodeset_compiler.html)) to easily register my structures and variables. But I would also like to use that XML file for type-safe code on the client, so that I could ensure at compile time that the OPC-UA variables I'm reading or writing are actually defined in the Nodeset, and so that I don't have to explicitly specify the variable types in my C++ code again (they are already specified in the Nodeset XML file after all).

And similarly I would like to have type-safe accessor methods on the server side, to read and write the variables that are provided by my server.

So far I did not find such a feature in the documentation. Is this possible with open62541/open62541pp? Or is this not intended, or out of scope? Or did I misunderstand OPC-UA in general, and this approach would not make sense at all?

Thanks, Oliver

**Comment by lukasberbuer** [comment]
Hi @oliver,

currently, there is no solution to generate type-safe C++ classes from XML - but I hope we find a suitable procedure for the future.
There was a discussion in #137 where I described two possible options.

I did some first experiments to generate classes from Python. To maintain interoperability with open62541, the generated types need to be describable by a `UA_DataType`.

Let's image a open62541-style C struct with all possible member types and how it is generated from a XML:

```c
struct ExampleStructC {
    // Primitive builtin type
    UA_Byte byte;
    // Non-primitive builtin type
    UA_String string;
    // Array
    size_t arrayDimensionsSize;
    UA_UInt32* arrayDimensions;
    // Optional
    UA_DiagnosticInfo* optionalInfo;
    // Union
    enum {
        UA_UNIONFIELD_ID = 0,
        UA_UNIONFIELD_NUMERIC = 1,
    } unionMemberSwitchField;
    union {
        UA_NodeId id;
        UA_Int32 numeric;
    } unionMember;
};
```

I imagine the generated C++ version should look like this:

```cpp
struct ExampleStructCpp {
    uint8_t byte;
    String string;
    Array<int32_t> arrayDimensions;
    Optional<DiagnosticInfo> optionalInfo;
    Union<NodeId, int32_t> unionMember;
};
```

We can't just use `std::vector` for `Array`, `std::optional` for `Optional` and `std::variant` for `Union` because the memory layout is an implementation detail and not standardized. But open62541 expects a concrete memory layout for all cases:
- arrays are just structs with a length (`size_t`) and a byte array (`UA_Byte*`)
- optional values are represented as pointers (`nullptr` if empty)
- unions are structs with a switch field (`enum`) and a C `union`

If we create corresponding RAII wrapper `Array<T>`, `Optional<T>` and `Union<T...>`, all custom types can be represented as structs.
Plus it will be way easier to create custom data types containing arrays of optionals manually.
Overall, it's quite some work to do, but we can break it down into small bits.

What's your take on this @oliver? How would you like the generated C++ types to look like?

**Comment by oliver** [comment]
Hi Lukas,

as mentioned, I'm not really experienced with OPC-UA or open62541(pp). And my Nodeset XML apparently does not have any custom types. So generating structs for custom types is far too advanced for me yet :-) and I cannot really comment how these generated structs should look.

At the moment I'm using open62541pp pretty much like in the [example code](https://github.com/open62541pp/open62541pp/blob/master/examples/client_minimal.cpp) (with the addition of a simple `find_namespace_index()` function):

```c++
const auto ns_index = find_namespace_index("http://www.siemens.com/simatic-s7-opcua", client.getNamespaceArray());
const opcua::NodeId node_id(ns_index, "SerialNumber");
auto node = client.getNode(node_id);
const auto value = node.readValueScalar<std::string>();
std::cout << "serial number: '" << value << "'" << std::endl;
```

I see two possible improvements here for type safety:
- generate `NodeId` objects also for variables from user-supplied nodesets (so that I can use them like the `opcua::VariableId::Server_ServerStatus_CurrentTime` constant in the example code)
	- I guess this is already done if I supply a .csv file with my desired IDs to the `generate_datatypes.py` script? Or is there another way that does not require me to maintain the .csv file manually?
- generate C++ type information for variables from user-supplied nodesets, so that I don't have to specify the data type when calling `readValueScalar()`.

My (naive) approach would look something like this:

```c++
const auto ns_index = find_namespace_index("http://www.siemens.com/simatic-s7-opcua", client.getNamespaceArray());
auto node = client.getTypedNode<plc::TypedVariableId::SerialNumber>(ns_index); // "node" will be of type "opcua::TypedScalarNode<std::string>"
const auto value = node.readValueScalar();
std::cout << "serial number: '" << value << "'" << std::endl;
```

I.e.:
- `plc::TypedVariableId::SerialNumber` is some generated C++ type that provides information about the node id and about the expected OPC-UA data type
- `client.getTypedNode()` makes use of that data type info to return a `TypedScalarNode<T>` object, which is of a template type, with the expected C++ data type as template parameter
- so when calling `client.getTypedNode()` or `TypedScalarNode<T>::readValueScalar()` I don't need to specify the data type any more

Does this make sense? Or am I misunderstanding something entirely here?

**Reply by lukasberbuer** [reply, marked as answer]
Ok, I think I understand your idea. You would like to bind pre-defined node ids to the variable type and enforce compile-time checks to read the correct type?
This approach is limited to variable (type) nodes but has to face the complexity of OPC UA:
- A variable node may have a data type. If not, any scalar or array data type can be written to this node (even by the client if the access level is set accordingly)
- A variable node may specify the value rank (any, scalar, array, x-dimensions or combinations). So a variable node may have scalar or array values
- ...

I don't see any generic way to enforce the binding of data types and variable nodes facing those issues.

### One idea for more compile-time checks

If you for example create a node with `opcua::VariableId::Server_ServerStatus_CurrentTime`, you know for sure that the node class will be a variable node. The `opcua::Node` class has no idea about the actual node class and exposes functionality for all node classes. But not all node classes support all actions are have all attributes: https://reference.opcfoundation.org/Core/Part3/v104/docs/5.9

`opcua::Node<T>` type could be extended to also contain the `NodeClass` if known:

```cpp
template <typename T, NodeClass Class = NodeClass::Unspecified>
opcua::Node;
```

`server.getNode(opcua::VariableId::Server_ServerStatus_CurrentTime)` should return a node of type `Node<Server, NodeClass::Variable>`. Further checks are then available to allow only certain invokations of methods.

**Reply by oliver** [reply]
OK, I am indeed not that familiar with the complexities of OPC-UA; thanks for pointing those out!

Your additional idea for compile-time checks looks interesting to me. I guess something like this would be useful; but I'll probably only be able to really evaluate this after I've finished my current OPC-UA project :-) . Still, it might be a worthwhile to put such a type-safe API on the "long-term improvements" list.
